### DISCUSSION TITLE
Magic Wand selection, different expectation in PD compared to other programs

### DISCUSSION THREAD

**Discussion by hi5** [tannerhelland/PhotoDemon#552] [Q&A] [answered]
Either it is the way it is e.g. how it should work, or I am missing something with regards to how to get the **Magic Wand** to work how I expect it would work.

In the image below there are three examples.

1. I have this single layer image, I click in the middle section and the Magic Wand selection selects what I more or less what I expect and I can change various setting to make it more or less flexible. Perfect, does what I expect.

2. But when I click in the Black Border -- in many images this will be a "Background color" if you will -- my expectation is that the black border would be selected as illustrated in image 3. But in PD it selects the _entire_ image. I've tried various settings but it always selects the entire image.

3. Illustration of what I would expect. This is also what happens in other graphics programs hence my expectation.

So any tips on to quickly select the black border in this example (here it is of course easy as I can use the rectangular selection tool, but with complicated shapes that is of course not an option. I just want to select all "continuous black" to do something with the "background" color.

![magickwand](https://github.com/tannerhelland/PhotoDemon/assets/1973899/3b90ab9b-6195-4bf9-af13-b5e68a5c132f)

**Comment by tannerhelland** [comment]
Hello @hi5 .  The behavior you describe is definitely not expected to me, either.  I'm sorry for the trouble!

I took a quick screenshot of your image (above) and cropped out image number 2 to test, and my local copy of PD selects just the black border as both of us would (I think) expect when using the magic wand tool:

![image](https://github.com/tannerhelland/PhotoDemon/assets/1930029/22eafe0f-cabf-43cb-99dd-b1aa1374527f)

The magic wand tool has a number of settings, so it's possible some combination produces unexpected results.  Could I ask for a quick follow-up from you to help me study the problem further?

1) A copy of the original image you used above (just in case it's some weird image format bug)
2) A screenshot of the "tolerance" panel of the magic wand tool, in case a specific combination of settings is what's broken.  Mine look like this (for the above screenshot):

![image](https://github.com/tannerhelland/PhotoDemon/assets/1930029/c4d37cdc-58e3-4176-9c1f-d56da60ad93b)

Thank you in advance.

**Comment by mizka76** [comment, marked as answer]
hello, tanner, & Hi5,
if I do remember well, I have seen this behaviour of the magic wand when antialiasing is on, ( select the whole image, not just the chosen color ) , but as antialiasing is set as "none" the selection is ok .
I have reproduced this here 
![magic wand   antialiasing (whole image)](https://github.com/tannerhelland/PhotoDemon/assets/62914748/9287a370-c391-414c-835f-9cedfb5c0edc)
![magic wand   no antialisaing (only background selected)](https://github.com/tannerhelland/PhotoDemon/assets/62914748/5beb4525-9d6c-43ff-aafd-f467ce9e25cd)

**Reply by tannerhelland** [reply]
Excellent detective work, @mizka76 !  Thank you so much for determining the cause of this bug.

I have uploaded a new nightly build with a fix for the antialiasing problem on magic wand selections.  Thank you again!

**Comment by hi5** [comment]
> hello, tanner, & Hi5, if I do remember well, I have seen this behaviour of the magic wand when antialiasing is on, ( select the whole image, not just the chosen color ) , but as antialiasing is set as "none" the selection is ok . 

Thank you, indeed setting **smoothing** it to **none** does the trick, probably the only thing I didn't try :)

My cool PNG test image just in case (created using PD btw)
![image](https://github.com/tannerhelland/PhotoDemon/assets/1973899/c1ef306d-f0af-4aeb-a235-f2695b006bb2)

**Reply by tannerhelland** [reply]
Antialiasing is also the only setting I failed to try!  Never fails 🤦‍♂️ 

I have uploaded a new nightly build with a fix for this issue, so antialiasing on magic wands should no longer be a problem.  (The bug was occurring with magic wand selections that have "holes" inside them; the holes were being mistakenly dropped by a path simplification algorithm that "smooths out" the edges of the magic wand selection prior to antialiasing, to improve the quality of the final result.)

Great find as usual, @hi5 .  Thank you so much for your ongoing help.  It means a lot.

**Reply by hi5** [reply]
Cheers, I can confirm the recent update now selects with all three smoothing settings. 👍
